sqlx-postgres/src/replication.rs

```diff
@@ -0,0 +1,244 @@
+                    MessageFormat::CopyDone => {
+                        let _ = message.decode::<CopyDone>()?;
+                        Some(PgCopyBothCommand::CopyDone { from_client: false })
+                    }
+                    _ => {
+                        return Err(err_protocol!(
+                            "unexpected message format during copy out: {:?}",
+                            message.format
+                        ))
+                    }
+                }
+            }
+            // This only errors if the consumer has been dropped.
+            // There is no reason to continue.
+            Either::Right((command, _truc)) => {
+                Some(command.or_else(|err| match err {
+                    flume::RecvError::Disconnected => {
+                        Ok(PgCopyBothCommand::CopyDone { from_client: true })
+                    }
+                    _ => Err(Error::WorkerCrashed),
+                })?)
+            }
+        };
+
+        if let Some(command) = command {
+            match command {
+                // Start the stream.
+                PgCopyBothCommand::Begin(buf) => {
+                    if has_started {
+                        return Err(err_protocol!("Copy-Both mode already initiated"));
+                    }
+                    conn.stream.send(buf.as_slice()).await?;
+                    // Consume the server response.
+                    conn.stream
+                        .recv_expect::<CopyResponse>(MessageFormat::CopyBothResponse)
+                        .await?;
+                    has_started = true;
+                }
+                // Send data to the server.
+                PgCopyBothCommand::CopyData(data) => {
+                    if !has_started {
+                        return Err(err_protocol!("connection hasn't been started"));
+                    }
+                    conn.stream.send(CopyData(data)).await?;
+                }
+
+                // Grafeceful shutdown of the stream.
+                PgCopyBothCommand::CopyDone { from_client } => {
+                    if !has_started {
+                        return Err(err_protocol!("connection hasn't been started"));
+                    }
+                    conn.stream.send(CopyDone).await?;
+                    // If we are the first to send CopyDone, wait for the server to send his own.
+                    if from_client {
+                        conn.stream.recv_expect(MessageFormat::CopyDone).await?;
+                    }
+                    conn.stream
+                        .recv_expect(MessageFormat::CommandComplete)
+                        .await?;
+                    conn.stream
+                        .recv_expect(MessageFormat::CommandComplete)
+                        .await?;
+                    conn.stream
+                        .recv_expect(MessageFormat::ReadyForQuery)
+                        .await?;
```

Comment:
An alternative to this diff is to call `conn.detach` instead. Any opinion @abonander?